drivers/sensor/npm1300_charger/npm1300_charger.c

```diff
@@ -0,0 +1,404 @@
+#define CHGR_OFFSET_CHG_STAT	0x34U
+#define CHGR_OFFSET_ERR_REASON	0x36U
+
+/* nPM1300 ADC register offsets */
+#define ADC_OFFSET_TASK_VBAT 0x00U
+#define ADC_OFFSET_TASK_TEMP 0x01U
+#define ADC_OFFSET_CONFIG    0x09U
+#define ADC_OFFSET_NTCR_SEL  0x0AU
+#define ADC_OFFSET_RESULTS   0x10U
+#define ADC_OFFSET_IBAT_EN   0x24U
+
+/* nPM1300 VBUS register offsets */
+#define VBUS_OFFSET_TASK_UPDATE 0x00U
+#define VBUS_OFFSET_ILIM	0x01U
+#define VBUS_OFFSET_STATUS	0x07U
+
+/* Ibat status */
+#define IBAT_STAT_DISCHARGE	 0x04U
+#define IBAT_STAT_CHARGE_TRICKLE 0x0CU
+#define IBAT_STAT_CHARGE_COOL	 0x0DU
+#define IBAT_STAT_CHARGE_NORMAL	 0x0FU
+
+struct adc_results_t {
+	uint8_t ibat_stat;
+	uint8_t msb_a[4];
+	uint8_t lsb_a;
+	uint8_t msb_b[4];
+	uint8_t lsb_b;
+} __packed;
+
+/* Linear range for charger terminal voltage */
+static const struct linear_range charger_volt_ranges[] = {
+	LINEAR_RANGE_INIT(3500000, 50000, 0U, 3U), LINEAR_RANGE_INIT(4000000, 50000, 4U, 13U)};
+
+/* Linear range for charger current */
+static const struct linear_range charger_current_range = LINEAR_RANGE_INIT(32000, 2000, 16U, 400U);
+
+/* Linear range for Discharge limit */
+static const struct linear_range discharge_limit_range = LINEAR_RANGE_INIT(268090, 3230, 83U, 415U);
+
+/* Linear range for vbusin current limit */
+static const struct linear_range vbus_current_ranges[] = {
+	LINEAR_RANGE_INIT(100000, 0, 1U, 1U), LINEAR_RANGE_INIT(500000, 100000, 5U, 15U)};
+
+/* Read multiple registers from specified address */
+static int reg_read_burst(const struct device *dev, uint8_t base, uint8_t offset, void *data,
+			  size_t len)
+{
+	const struct npm1300_charger_config *const config = dev->config;
+	uint8_t buff[] = {base, offset};
+
+	return i2c_write_read_dt(&config->i2c, buff, sizeof(buff), data, len);
+}
+
+static int reg_read(const struct device *dev, uint8_t base, uint8_t offset, uint8_t *data)
+{
+	return reg_read_burst(dev, base, offset, data, 1U);
+}
+
+/* Write single register to specified address */
+static int reg_write(const struct device *dev, uint8_t base, uint8_t offset, uint8_t data)
+{
+	const struct npm1300_charger_config *const config = dev->config;
+	uint8_t buff[] = {base, offset, data};
+
+	return i2c_write_dt(&config->i2c, buff, sizeof(buff));
+}
+
+static int reg_write2(const struct device *dev, uint8_t base, uint8_t offset, uint8_t data1,
+		      uint8_t data2)
+{
+	const struct npm1300_charger_config *const config = dev->config;
+	uint8_t buff[] = {base, offset, data1, data2};
+
+	return i2c_write_dt(&config->i2c, buff, sizeof(buff));
+}
+
+static void calc_temp(const struct npm1300_charger_config *const config, uint16_t code,
+		      struct sensor_value *valp)
+{
+	float log_result = log((1024.f / (float)code) - 1);
+	float inv_temp_k = (1.f / 298.15f) - (log_result / (float)config->thermistor_beta);
+
+	float temp = (1.f / inv_temp_k) - 273.15f;
+
+	valp->val1 = (int32_t)temp;
+	valp->val2 = (int32_t)(fmodf(temp, 1.f) * 1000000.f);
```

Comment:
is `fmodf` the reason for newlib later? If so, I'd avoid using it. Actually, integer operations are in general preferred even though not mandatory.

Reply:
@stephanosio is there an alternative with minimal libc or picolibc for this? and does it make sense?

Reply:
> To avoid this it would be necessary to pre-calculate a lookup table for the selected thermistor.

I would seriously consider this to be honest in this case.

Reply:
`arm_vlog_*` from CMSIS-DSP may be an option? Is linearization (or some sort of approximation) possible here?

Reply:
> @stephanosio is there an alternative with minimal libc or picolibc for this? and does it make sense?

Picolibc, or any other "full" libc for that matter, has `fmodf`.

If you **must** use `fmodf`, you can `select REQUIRES_FULL_LIBC` from `config NPM1300_CHARGER` to indicate that this driver requires a "full" libc (e.g. picolibc, newlib, armclang libc, ...).

By the way, this will be a non-issue once we make Picolibc the default libc in the future (after all, this type of issues is one of the reasons why we are doing that in the first place).

Reply:
@carlescufi I have replaced the floating point calculations with a pre-calculated lookup table, and a script to generate it.

Reply:
After looking into the implementation, I think the initial version was better. In the end we're complicating things because the minimal libc doesn't support any math function (except sqrt/f). AFAIK, we'll soon transition to picolibc, so we'll have the full math.h. So, I'd just make the driver `select REQUIRES_FULL_LIBC`. Another option would be to add `logf` (and its dependencies) to minimal libc, https://github.com/apache/nuttx/blob/master/libs/libm/libm/lib_logf.c could be an option, looks like it's MIT licensed @stephanosio ?